### PR TITLE
feat: added use_internal_docker_network feature

### PR DESCRIPTION
**Pull request by amalucelli**
As I don't have write access in the base branch of the PR #283, I forked from `master`, made the changes of the origin PR and also the one suggested in the comments.

Co-authored-by: @AaronKalair

### CODE REVIEW

**Comment by jcalonso** [comment]
+1 for this, I'm running a Dind (Docker in Docker) setup and I'm getting the following error:
```
----End Backtrace-----
[integration:L554:28s] D      ---Nested Exception---
[integration:L555:28s] D      Class: Kitchen::ActionFailed
[integration:L556:28s] D      Message: Failed to complete #create action: [Address not available - connect(2) for [::1]:32771]
[integration:L557:28s] D      ----------------------
[integration:L558:28s] D      ------Backtrace-------
[integration:L559:28s] D      /usr/lib/ruby/2.4.0/socket.rb:1198:in `__connect_nonblock'
[integration:L560:28s] D      /usr/lib/ruby/2.4.0/socket.rb:1198:in `connect_nonblock'
[integration:L561:28s] D      /usr/lib/ruby/2.4.0/socket.rb:56:in `connect_internal'
[integration:L562:28s] D      /usr/lib/ruby/2.4.0/socket.rb:137:in `connect'
[integration:L563:28s] D      /usr/lib/ruby/2.4.0/socket.rb:627:in `block in tcp'
[integration:L564:28s] D      /usr/lib/ruby/2.4.0/socket.rb:227:in `each'
[integration:L565:28s] D      /usr/lib/ruby/2.4.0/socket.rb:227:in `foreach'
[integration:L566:28s] D      /usr/lib/ruby/2.4.0/socket.rb:617:in `tcp'
[integration:L567:28s] D      /usr/lib/ruby/gems/2.4.0/gems/net-ssh-4.2.0/lib/net/ssh/transport/session.rb:70:in `initialize'
[integration:L568:28s] D      /usr/lib/ruby/gems/2.4.0/gems/net-ssh-4.2.0/lib/net/ssh.rb:237:in `new'
[integration:L569:28s] D      /usr/lib/ruby/gems/2.4.0/gems/net-ssh-4.2.0/lib/net/ssh.rb:237:in `start'
...
```

**Comment by robertlabrie** [comment]
Can we please have this? I do a terrible hack of clobbering `localhost` in /etc/hosts to work around it

File: lib/kitchen/driver/docker.rb [hunk 1]
```diff
@@ -384,6 +393,17 @@ def container_ssh_port(state)
         end
       end
 
+      def container_ip(state)
+        begin
+          cmd = "inspect --format '{{ .NetworkSettings.IPAddress }}'"
```
Comment (rs017991) [inline lib/kitchen/driver/docker.rb]: @amalucelli @AaronKalair FYI, there is an issue when using this setup with user defined networks.

In those cases, the IPAddress at the NetworkSettings level is empty, and instead needs to be found on the Network. That nested value is also there when using the default bridge network, so I would think that it should be safe to use in any case:
I was able to get this working by changing the template to `'{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}'`

**Comment by timogoebel** [comment]
Nice, I needed this to get `kitchen-docker` working on gitlab-ci.

Please see the `.gitlab-ci.yml` on how to use this patch before it's merged.

```yaml
image: ruby:2.5-alpine
services:
  - docker:dind

stages:
  - functional

verify-centos-7:
  before_script:
    - apk add docker libffi-dev build-base patch
    - docker info
    - bundle install
    - cd $(bundle show kitchen-docker)
    - wget -O use_internal_docker_network.patch https://patch-diff.githubusercontent.com/raw/test-kitchen/kitchen-docker/pull/304.patch
    - patch -p1 < use_internal_docker_network.patch
    - cd -
  stage: functional
  tags:
    - aws
  artifacts:
    expire_in: 7d
    when: on_failure
    paths:
      - .kitchen/logs
  script:
    - kitchen test default-centos-7 -d always
```

**Comment by ayanich** [comment]
I also need this to get things running on circleci, would love to see this get merged.

**Comment by jtimberman** [comment]
Hello,

I've verified that this works using our CI system, too.

This script is what I used to apply the PR, install the gem with ChefDK, and run Test Kitchen in my cookbook repository.

```sh
#!/bin/bash
mkdir -p /kitchen
pushd /kitchen
  git clone https://github.com/test-kitchen/kitchen-docker
  pushd kitchen-docker
  git fetch origin pull/304/head:use_internal_docker_network
  git checkout use_internal_docker_network
  chef exec bundle install
  chef exec bundle exec rake install
  popd
popd

chef exec kitchen test
```

**Comment by tas50** [comment]
Merging based on @jtimberman's experience

**Comment by Nklya** [comment]
It seems that I cannot use kitchen-docker with `use_internal_docker_network: true` on macOS because docker for mac doesn't connect local and docker networks.

Will investigate and maybe create another PR.

**Pull request merged**